docs/tla/ThreePhase.tla

```diff
@@ -0,0 +1,111 @@
+            /\ store = Store
+
+TPConsistent == \A r1, r2 \in RM : ~ /\ rmState[r1] = "aborted"
+                                     /\ rmState[r2] = "pre-committed"
+
+TMRcvPrepared(t, r) == /\ tmState[t] = "init"
+                    /\ [type |-> "Prepared", rm |-> r] \in msgs
+                    /\ tmPrepared' = tmPrepared \union {r}
+                    /\ UNCHANGED << rmState, tmState, msgs >>                                  
+
+TMPreCommit(t) == /\ tmState[t] = "init"
+            /\ tmState' = "prepared"
+            /\ tmPrepared = RM
+            /\ msgs' =  msgs \union {[type |-> "PreCommit"]}
+            /\ UNCHANGED << rmState, tmPrepared, store >>  
+
+TMCommit(t) == /\ tmState[t] = "prepared"
+            /\ tmState' = "done"
+            /\ tmPrepared = RM
+            /\ msgs' =  msgs \union {[type |-> "Commit"]}
+            /\ UNCHANGED << rmState, tmPrepared, store >>
+
+TMAbort(t) ==  /\ tmState[t] = "init"
+            /\ tmState' = "aborted"
+            /\ msgs' =  msgs \union {[type |-> "Abort"]}
+            /\ UNCHANGED << rmState, tmPrepared, store >>
+           
+RMPrepare(r) == /\ rmState[r] = "working"
+                /\ rmState' = [rmState EXCEPT ![r] = "prepared"]
+                /\ msgs' = msgs \union {[type |-> "Prepared", rm |-> r]}
+                /\ UNCHANGED << tmState, tmPrepared, store >>
+
+RMChooseToAbort(r) == \* /\ \E t \in TM: store[t][r] \notin {NoVal, r}                  
+                      /\ rmState[r] \in {"working", "prepared"}
+                      /\ rmState' = [rmState EXCEPT ![r] = "aborted"]
+                      /\ UNCHANGED << tmState, tmPrepared, msgs, store >>  
+
+RMRcvPreCommitMsg(r) == \* /\ \E t \in TM : store[t][r] \* Connected
+                        /\ [type |-> "PreCommit"] \in msgs
+                        /\ rmState' = [rmState EXCEPT ![r] = "pre-committed"]
+                        /\ UNCHANGED << tmState, tmPrepared, msgs, store >> 
+
+RMRcvCommitMsg(r) == \* /\ \E t \in TM : store[t][r] \* Connected
+                     /\ [type |-> "Commit"] \in msgs
+                     /\ rmState' = [rmState EXCEPT ![r] = "committed"]
+                     /\ UNCHANGED << tmState, tmPrepared, msgs >> 
+ 
+RMRcvAbortMsg(r) == \* /\ \E t \in TM : store[t][r] \* Connected
+                    /\ [type |-> "Abort"] \in msgs
+                    /\ rmState' = [rmState EXCEPT ![r] = "aborted"]
+                    /\ UNCHANGED << tmState, tmPrepared, msgs >>
+
+RMunavailable(t,r) == /\ rmState[r] = "prepared"
+                      /\ store[t][r] /= {}
+                      /\ store' = [store EXCEPT ![t][r] = {}]
+                      /\ UNCHANGED << tmState, tmPrepared, msgs >>
+
+TPNext == \/ \E t \in TM : \E r \in RM :
+            \/ TMCommit(t) \/ TMPreCommit(t) \/ TMAbort(t)
+            \/ TMRcvPrepared(t, r) \/ RMPrepare(r) \/ RMChooseToAbort(r) 
+            \/ RMRcvPreCommitMsg(r) \/ RMRcvCommitMsg(r) \/ RMRcvAbortMsg(r)
+
+\* ---------------
+INSTANCE TCommit
```

Comment:
It would be reasonable to inherit the consistency theorem from transaction commit specification.

Reply:
File is removed.